File: langs/python/forgeh/src/forgeh/forgemix.py
"""ForgeHash-B3 v1 ForgeMix compression (SPECIFICATION.md §14)."""

from __future__ import annotations

import struct
from typing import List, Optional

BLOCK_SIZE = 1024
WORDS = 128
ROUNDS = 8

MASK64 = (1 << 64) - 1
MASK32 = 0xFFFFFFFF

WordBlock = List[int]


def low32(x: int) -> int:
    return x & MASK32


def rotl(x: int, n: int) -> int:
    n &= 63
    return ((x << n) | (x >> (64 - n))) & MASK64


def rotr(x: int, n: int) -> int:
    n &= 63
    return ((x >> n) | (x << (64 - n))) & MASK64


def quarter_round(words: WordBlock, base: int, i0: int, i1: int, i2: int, i3: int) -> None:
    a = words[base + i0]
    b = words[base + i1]
    c = words[base + i2]
    d = words[base + i3]

    a = (a + b + 2 * low32(a) * low32(b)) & MASK64
    d = rotr(d ^ a, 32)

    c = (c + d + 2 * low32(c) * low32(d)) & MASK64
    b = rotr(b ^ c, 24)

    a = (a + b + 2 * low32(a) * low32(b)) & MASK64
    d = rotr(d ^ a, 16)

    c = (c + d + 2 * low32(c) * low32(d)) & MASK64
    b = rotr(b ^ c, 63)

    words[base + i0] = a
    words[base + i1] = b
    words[base + i2] = c
    words[base + i3] = d


def apply_schedule(words: WordBlock, base: int) -> None:
    quarter_round(words, base, 0, 4, 8, 12)
    quarter_round(words, base, 1, 5, 9, 13)
    quarter_round(words, base, 2, 6, 10, 14)
    quarter_round(words, base, 3, 7, 11, 15)
    quarter_round(words, base, 0, 5, 10, 15)
    quarter_round(words, base, 1, 6, 11, 12)
    quarter_round(words, base, 2, 7, 8, 13)
    quarter_round(words, base, 3, 4, 9, 14)


def _idx(row: int, col: int) -> int:
    return row * 16 + col


def mix_rows(state: WordBlock) -> None:
    for row in range(8):
        apply_schedule(state, row * 16)


_virtual_row = [0] * 16
_group = [0] * 16


def mix_columns(state: WordBlock) -> None:
    for pair in range(8):
        col_a = pair * 2
        col_b = col_a + 1
        for row in range(8):
            _virtual_row[row * 2] = state[_idx(row, col_a)]
            _virtual_row[row * 2 + 1] = state[_idx(row, col_b)]
        apply_schedule(_virtual_row, 0)
        for row in range(8):
            state[_idx(row, col_a)] = _virtual_row[row * 2]
            state[_idx(row, col_b)] = _virtual_row[row * 2 + 1]


def mix_diagonals(state: WordBlock) -> None:
    for diagonal_index in range(8):
        base = diagonal_index * 2
        for k in range(8):
            _group[k * 2] = state[_idx(k, (base + k) & 15)]
            _group[k * 2 + 1] = state[_idx(k, (base + k + 8) & 15)]
        apply_schedule(_group, 0)
        for k in range(8):
            state[_idx(k, (base + k) & 15)] = _group[k * 2]
            state[_idx(k, (base + k + 8) & 15)] = _group[k * 2 + 1]


def permute(state: WordBlock, permuted: WordBlock) -> None:
    for source in range(WORDS):
        dest = (source * 73 + 19) & 127
        permuted[dest] = state[source]


def mix(
    input_block: WordBlock,
    pass_: int,
    lane: int,
    block_index: int,
    output: Optional[WordBlock] = None,
) -> WordBlock:
    """ForgeMix(inputBlock, pass, lane, blockIndex) -> outputBlock."""
    original = input_block
    state = list(input_block)

    state[0] ^= pass_ & MASK64
    state[1] ^= lane & MASK64
    state[2] ^= block_index & MASK64
    state[3] ^= rotl((pass_ + block_index) & MASK64, 17)

    permuted = [0] * WORDS
    for _ in range(ROUNDS):
        mix_rows(state)
        mix_columns(state)
        mix_diagonals(state)
        permute(state, permuted)
        state, permuted = permuted, state

    out = output if output is not None else [0] * WORDS
    for i in range(WORDS):
        out[i] = state[i] ^ original[i]
    return out


def bytes_to_words(data: bytes | bytearray | memoryview, out: Optional[WordBlock] = None) -> WordBlock:
    words = out if out is not None else [0] * WORDS
    unpacked = struct.unpack_from("<128Q", data)
    for i in range(WORDS):
        words[i] = unpacked[i]
    return words


def words_to_bytes(words: WordBlock, out: Optional[bytearray] = None) -> bytes | bytearray:
    packed = struct.pack("<128Q", *[w & MASK64 for w in words])
    if out is None:
        return packed
    out[:BLOCK_SIZE] = packed
    return out


def xor_words(left: WordBlock, right: WordBlock, out: Optional[WordBlock] = None) -> WordBlock:
    dest = out if out is not None else [0] * WORDS
    for i in range(WORDS):
        dest[i] = left[i] ^ right[i]
    return dest

File: src/ForgeHash.Core/ForgeMix.cs
using System.Buffers.Binary;
using System.Runtime.CompilerServices;

namespace ForgeHash;

/// <summary>
/// ForgeMix compression function: eight full rounds of row, column, and diagonal mixing
/// with a fixed word permutation and XOR feed-forward.
/// </summary>
public static class ForgeMix
{
    /// <summary>Block size in bytes.</summary>
    public const int BlockSize = 1024;

    /// <summary>Words per block.</summary>
    public const int WordsPerBlock = 128;

    /// <summary>Number of full rounds.</summary>
    public const int RoundCount = 8;

    /// <summary>
    /// Compresses a 1024-byte input block with positional metadata into a 1024-byte output block.
    /// </summary>
    public static void Mix(
        ReadOnlySpan<ulong> inputBlock,
        ulong pass,
        ulong lane,
        ulong blockIndex,
        Span<ulong> outputBlock)
    {
        if (inputBlock.Length != WordsPerBlock)
        {
            throw new ArgumentException("Input block must contain 128 words.", nameof(inputBlock));
        }

        if (outputBlock.Length != WordsPerBlock)
        {
            throw new ArgumentException("Output block must contain 128 words.", nameof(outputBlock));
        }

        Span<ulong> state = stackalloc ulong[WordsPerBlock];
        Span<ulong> original = stackalloc ulong[WordsPerBlock];
        Span<ulong> permuted = stackalloc ulong[WordsPerBlock];

        inputBlock.CopyTo(original);
        inputBlock.CopyTo(state);

        InjectPosition(state, pass, lane, blockIndex);

        for (int round = 0; round < RoundCount; round++)
        {
            MixRows(state);
            MixColumns(state);
            MixDiagonals(state);
            PermuteWords(state, permuted);
            permuted.CopyTo(state);
        }

        unchecked
        {
            for (int i = 0; i < WordsPerBlock; i++)
            {
                outputBlock[i] = state[i] ^ original[i];
            }
        }
    }

    /// <summary>
    /// Byte-oriented convenience wrapper that interprets blocks as little-endian UInt64 words.
    /// </summary>
    public static void Mix(
        ReadOnlySpan<byte> inputBlock,
        ulong pass,
        ulong lane,
        ulong blockIndex,
        Span<byte> outputBlock)
    {
        if (inputBlock.Length != BlockSize || outputBlock.Length != BlockSize)
        {
            throw new ArgumentException("Blocks must be exactly 1024 bytes.");
        }

        Span<ulong> inputWords = stackalloc ulong[WordsPerBlock];
        Span<ulong> outputWords = stackalloc ulong[WordsPerBlock];
        BytesToWords(inputBlock, inputWords);
        Mix(inputWords, pass, lane, blockIndex, outputWords);
        WordsToBytes(outputWords, outputBlock);
    }

    /// <summary>
    /// Quarter-round used by row, column, and diagonal mixing.
    /// </summary>
    public static void QuarterRound(ref ulong a, ref ulong b, ref ulong c, ref ulong d)
    {
        unchecked
        {
            a = a + b + (2UL * Low32(a) * Low32(b));
            d = RotateRight64(d ^ a, 32);

            c = c + d + (2UL * Low32(c) * Low32(d));
            b = RotateRight64(b ^ c, 24);

            a = a + b + (2UL * Low32(a) * Low32(b));
            d = RotateRight64(d ^ a, 16);

            c = c + d + (2UL * Low32(c) * Low32(d));
            b = RotateRight64(b ^ c, 63);
        }
    }

    private static void InjectPosition(Span<ulong> state, ulong pass, ulong lane, ulong blockIndex)
    {
        unchecked
        {
            state[0] ^= pass;
            state[1] ^= lane;
            state[2] ^= blockIndex;
            state[3] ^= RotateLeft64(pass + blockIndex, 17);
        }
    }

    private static void MixRows(Span<ulong> state)
    {
        for (int row = 0; row < 8; row++)
        {
            int baseIndex = row * 16;
            ApplySixteenWordSchedule(state, baseIndex);
        }
    }

    private static void MixColumns(Span<ulong> state)
    {
        Span<ulong> virtualRow = stackalloc ulong[16];

        for (int pair = 0; pair < 8; pair++)
        {
            int colA = pair * 2;
            int colB = colA + 1;

            for (int row = 0; row < 8; row++)
            {
                virtualRow[row * 2] = state[Index(row, colA)];
                virtualRow[row * 2 + 1] = state[Index(row, colB)];
            }

            ApplySixteenWordSchedule(virtualRow, 0);

            for (int row = 0; row < 8; row++)
            {
                state[Index(row, colA)] = virtualRow[row * 2];
                state[Index(row, colB)] = virtualRow[row * 2 + 1];
            }
        }
    }

    private static void MixDiagonals(Span<ulong> state)
    {
        Span<ulong> group = stackalloc ulong[16];

        for (int diagonalIndex = 0; diagonalIndex < 8; diagonalIndex++)
        {
            int @base = diagonalIndex * 2;

            for (int k = 0; k < 8; k++)
            {
                group[k * 2] = state[Index(k, (@base + k) & 15)];
                group[k * 2 + 1] = state[Index(k, (@base + k + 8) & 15)];
            }

            ApplySixteenWordSchedule(group, 0);

            for (int k = 0; k < 8; k++)
            {
                state[Index(k, (@base + k) & 15)] = group[k * 2];
                state[Index(k, (@base + k + 8) & 15)] = group[k * 2 + 1];
            }
        }
    }

    /// <summary>
    /// Applies the fixed eight quarter-round schedule used for every 16-word group.
    /// </summary>
    private static void ApplySixteenWordSchedule(Span<ulong> words, int offset)
    {
        QuarterRound(ref words[offset + 0], ref words[offset + 4], ref words[offset + 8], ref words[offset + 12]);
        QuarterRound(ref words[offset + 1], ref words[offset + 5], ref words[offset + 9], ref words[offset + 13]);
        QuarterRound(ref words[offset + 2], ref words[offset + 6], ref words[offset + 10], ref words[offset + 14]);
        QuarterRound(ref words[offset + 3], ref words[offset + 7], ref words[offset + 11], ref words[offset + 15]);

        QuarterRound(ref words[offset + 0], ref words[offset + 5], ref words[offset + 10], ref words[offset + 15]);
        QuarterRound(ref words[offset + 1], ref words[offset + 6], ref words[offset + 11], ref words[offset + 12]);
        QuarterRound(ref words[offset + 2], ref words[offset + 7], ref words[offset + 8], ref words[offset + 13]);
        QuarterRound(ref words[offset + 3], ref words[offset + 4], ref words[offset + 9], ref words[offset + 14]);
    }

    private static void PermuteWords(ReadOnlySpan<ulong> state, Span<ulong> permuted)
    {
        for (int sourceIndex = 0; sourceIndex < WordsPerBlock; sourceIndex++)
        {
            int destinationIndex = ((sourceIndex * 73) + 19) & 127;
            permuted[destinationIndex] = state[sourceIndex];
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static int Index(int row, int column) => (row * 16) + column;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static ulong Low32(ulong value) => value & 0xFFFF_FFFFUL;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static ulong RotateRight64(ulong value, int bits)
        => (value >> bits) | (value << (64 - bits));

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static ulong RotateLeft64(ulong value, int bits)
        => (value << bits) | (value >> (64 - bits));

    internal static void BytesToWords(ReadOnlySpan<byte> bytes, Span<ulong> words)
    {
        for (int i = 0; i < WordsPerBlock; i++)
        {
            words[i] = BinaryPrimitives.ReadUInt64LittleEndian(bytes.Slice(i * 8, 8));
        }
    }

    internal static void WordsToBytes(ReadOnlySpan<ulong> words, Span<byte> bytes)
    {
        for (int i = 0; i < WordsPerBlock; i++)
        {
            BinaryPrimitives.WriteUInt64LittleEndian(bytes.Slice(i * 8, 8), words[i]);
        }
    }
}
